src/mpid/ch4/src/ch4_coll_impl.h

```diff
@@ -416,6 +419,112 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_Bcast_intra_composition_gamma(void *buffer, M
     goto fn_exit;
 }
 
+/*
+ * This type of composition performs well for GPU bcast as it can utilize the direct links between
+ * the GPUs in the same node. It has four steps:
+ * 1. Root copies data to the host.
+ * 2. Inter-node bcast among the leaders on the host.
+ * 3. Leaders copy data to GPU.
+ * 4. Intra-node bcast in each node on the GPU buffer.
+*/
+MPL_STATIC_INLINE_PREFIX int MPIDI_Bcast_intra_composition_delta(void *buffer, MPI_Aint count,
+                                                                 MPI_Datatype datatype,
+                                                                 int root, MPIR_Comm * comm,
+                                                                 MPIR_Errflag_t errflag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int coll_ret = MPI_SUCCESS;
+    void *host_buffer = NULL;
+    void *saved_buffer = buffer;
+    MPL_pointer_attr_t attr;
+    MPI_Aint size, shift;
+
+#ifdef HAVE_ERROR_CHECKING
+    MPI_Status status;
+    MPI_Aint nbytes, recvd_size, type_size;
+#endif
+
+    if (comm->node_roots_comm == NULL && comm->rank == root) {
+        coll_ret = MPIC_Send(buffer, count, datatype, 0, MPIR_BCAST_TAG, comm->node_comm, errflag);
+        MPIR_ERR_COLL_CHECKANDCONT(coll_ret, errflag, mpi_errno);
+    }
+
+    if (comm->node_roots_comm != NULL && comm->rank != root &&
+        MPIR_Get_intranode_rank(comm, root) != -1) {
+#ifndef HAVE_ERROR_CHECKING
+        coll_ret =
+            MPIC_Recv(buffer, count, datatype, MPIR_Get_intranode_rank(comm, root), MPIR_BCAST_TAG,
+                      comm->node_comm, MPI_STATUS_IGNORE);
+        MPIR_ERR_COLL_CHECKANDCONT(coll_ret, errflag, mpi_errno);
+#else
+        coll_ret =
+            MPIC_Recv(buffer, count, datatype, MPIR_Get_intranode_rank(comm, root), MPIR_BCAST_TAG,
+                      comm->node_comm, &status);
+        MPIR_ERR_COLL_CHECKANDCONT(coll_ret, errflag, mpi_errno);
+
+        MPIR_Datatype_get_size_macro(datatype, type_size);
+        nbytes = type_size * count;
+        /* check that we received as much as we expected */
+        MPIR_Get_count_impl(&status, MPI_BYTE, &recvd_size);
+        if (recvd_size != nbytes) {
+            MPIR_ERR_SET2(coll_ret, MPI_ERR_OTHER,
+                          "**collective_size_mismatch",
+                          "**collective_size_mismatch %d %d", recvd_size, nbytes);
+            MPIR_ERR_COLL_CHECKANDCONT(coll_ret, errflag, mpi_errno);
+        }
+#endif
+    }
+
```

Comment:
Group the next two blocks of code under 
```
if (comm->node_roots_comm) {
    /* bcast in node_roots_comm */
    int inter_root = MPIR_Get_internode_rank(comm, root);
    int my_rank = comm->node_roots_comm->rank;
    if (my_rank == inter_root) {
        ...
    } else {
        ...
    }
}
```
But shouldn't `MPIDI_NM_mpi_bcast` take care of buffer swap anyway?

Reply:
Do you mean group the buffer allocation and inter-node bcast?
I can try to use MPIDI_NM_mpi_bcast to take care of the buffer swap, the performance should be similar.

Reply:
I think doing the explicit data swap as the current code is better than using MPIDI_NM_mpi_bcast to take care of the buffer swap as it is easier to see the difference between the composition alpha and delta.